### DISCUSSION TITLE
Zero-code adoption using drop-in runtimes

### DISCUSSION THREAD

**Discussion by shykes** [dagger/cloak#17] [Ideas]
## Overview

The design of Cloak opens the possibility of *zero-code adoption*, where a new user gets value from Dagger without having to write any code. That would make Dagger much easier to learn and adopt, which would greatly help with adoption.

This would be made possible by 1) implementing a system of modular runtimes, and 2) developing "drop-in runtimes" capable of natively running a project's pre-existing automation without modification, including:
  * Dockerfiles
  * Docker Compose files
  * Makefiles
  * NPM scripts
  * etc.

## Onboarding path

The complete onboarding path could look something like this:

1. Drop in. "Dagger runs my existing automation, better than the original tools (faster, more secure, more scalable, more portable)
2. Extend. "I can extend my existing automation with new actions, using a Dagger SDK for my favorite language. The developer experience is better, and I don't lose compatibility"
3. Refactor. "My team needs better tooling. I will replace our artisanal scripts with a custom tool written in a modern language. I can do this on top of the Dagger SDK, and save time by reusing all the actions I've already written so far.
4. Modularize. "My teams needs to customize their automation without re-writing if from scratch. I show them how to use Dagger to consume my automations as an API, and create their own custom automations on top, using the language of their choice.

## Design considerations

I’m assuming these runtimes will need to implement a special hook (itself running on dagger of course) that takes a source repo as input, and outputs a graphql schema exposing all the actions, artifacts and services that it found.

```mermaid
graph BT;
  project[Project Directory];
  subgraph "Dagger"
  cloak(["Cloak engine"])
  go("Go runtime");
  typescript("Typescript runtime");
  dockerfile("Dockerfile runtime");
  dockercompose("Docker Compose runtime");
  make("Make runtime");
  yarn("Yarn runtime");
  cloak --- go & typescript & dockerfile & dockercompose & make & yarn;
  end;
  project --- cloak;

    subgraph "Actions";
    test("test");
    lint("lint");
    deploy("deploy")
   end;

    subgraph "Services";
    www("www");
    api("api");
    db("db");
    end;

    subgraph "Artifacts";
    build("yarn/build")
    head("git/head")
    checkout("git/checkout")
    container("dockerfile/stage/main")
    buildenv("dockerfile/stage/buildenv")
    end;

 go --- deploy;
 make --- lint;
 yarn --- test & build;
 dockercompose --- www & api & db;
 dockerfile --- container & buildenv;
 
 class project secondary;
```

Presumably the Go & Typescript SDKs (which do require coding to use, by definition) would ship with a runtime of their own, implementing the same hooks. Perhaps their implementation will be very simple (if there is no boilerplate to generate) or become more advanced (if code and/or schema generation is involved). This matches common use of the terms “runtime” and “sdk”, as in “the JDK vX ships with the JRE vY”.

**Comment by aluzzardi** [comment]
Definitely interesting.

/cc @marcosnils -- we were briefly discussing no/low code adoption yesterday, this approach could be it (actions are available to run before you write a single line of code).

I think the gap to make this possible is smaller than it appears. I don't think we would need to generate schemas on the fly, GraphQL is flexible enough that a static schema works. For instance for `make`:

```graphql
# List the targets
make {
  targets {
    name
  }
}

# Run `make test`
make {
  target(name: "test") {
    run {
      stdout
    }
  }
}
```

**Reply by shykes** [reply]
One downside of a static schema is that the generated client stubs for them will be less useful (because the schema contains less information about what actions are available). So a "build" action implemented as a yarn script would not be a first-class citizen compared to one implemented in a native SDK: no auto-complete from the IDE, no built-time type checking etc.

wdyt?

**Reply by sipsma** [reply]
> One downside of a static schema is that the generated client stubs for them will be less useful

I was going to say the same thing, but I think this doesn't have to be a binary choice. We could support runtimes supplying a static schema (in cases where that makes sense, or in this Makefile example, maybe that's just what's practical for a v1 implementation even if the more typed version would be preferable). But we could also support them providing dynamically generated schemas too. Both seem entirely possible to support in a pretty cohesive way. Dagger just asks the runtime for the schema, whether its dynamic or static is an implementation detail of the runtime (or something like that).

**Reply by shykes** [reply]
Yes that's true. As long as the architecture allows for both, we can let the community experiment with both.

**Reply by sipsma** [reply]
Addendum to previous comment: Actually if we have a totally generic untyped API then we probably wouldn't need a runtime at all; it could just fit in the current action model today. I could already just go add a `make` action written in Go or TS that accepts as input a filesystem and a target to run. So maybe viewing this as a custom runtime only makes sense in the case where we are dynamically generating a schema derived from the makefile? Not sure yet, worth thinking through this distinction more

**Reply by shykes** [reply]
Yes I think that’s right. I’m interested in the runtime + dynamic schema use case :)

**Reply by aluzzardi** [reply]
> One downside of a static schema is that the generated client stubs for them will be less useful (because the schema contains less information about what actions are available). So a "build" action implemented as a yarn script would not be a first-class citizen compared to one implemented in a native SDK: no auto-complete from the IDE, no built-time type checking etc.

That's the typical GraphQL pattern, to have static schemas (e.g. you'd query `{ users(name: "aluzzardi") { email } }` rather than `{ aluzzardi { email } }`.

Having a dynamic schema prevents to programmatically configure requests (unless you sprintf your query which is an anti pattern), rather than binding variables.

For bindings, doing things like `for _, target := range make.Targets() { make.Run(target) }` becomes complicated (unless you use reflection to figure out what's available).

**Reply by sipsma** [reply]
> For bindings, doing things like for _, target := range make.Targets() { make.Run(target) } becomes complicated (unless you use reflection to figure out what's available).

I think the way this would work is that:
1. You are writing an action that relies on some Makefile, so you add a dependency (to cloak.yaml or whatever replaces it) that points to the fs containing the Makefile and specifies the `make` runtime
2. You then run `cloak generate` which introspects the makefile and dynamically converts it into a graphql schema. That enables generation of typed client stubs for the language you are writing in
3. You then code directly against that typed schema to call the makefile targets (and set args, etc.)

I think there's plenty of questions coming out of this (e.g. is it really annoying to have to regenerate client stubs every time the makefile changes, probably depends on the use case?), but it does seem to align with the overall pattern of how we generate client stubs today with go/ts action schemas.

**Reply by shykes** [reply]
> That's the typical GraphQL pattern, to have static schemas (e.g. you'd query `{ users(name: "aluzzardi") { email } }` rather than `{ aluzzardi { email } }`.

But if I implement a build action, I want to write `func Build()`, not `func Action(name string) { if name == "build" { .. } }`, right? And I want my action to be invoked by client with `foo.Build()`, rather than `foo.Do("build")`? What I'm saying is, if we want that pattern for actions written in Go, we would want the same patterns for actions written for other platforms. Otherwise, the magic of "zero-code adoption" is broken IMO.

**Reply by shykes** [reply]
Maybe we are not talking about the same thing with the term "dynamic schema". I'm talking about having a middleware that produces a schema for cloak to consume. Whether that schema is dynamically generated is up to the middleware. Once generated and loaded by cloak, it is "static" and therefore would require no particular graphql magic, and it would not complicate the usage patterns.

**Reply by aluzzardi** [reply]
> Addendum to previous comment: Actually if we have a totally generic untyped API then we probably wouldn't need a runtime at all; it could just fit in the current action model today. 

Yes, that's what I was thinking. I think we can get reach mostly the same outcome with the action model without an extra layer

**Reply by shykes** [reply]
I think having regular `make` and `yarn` actions as you describe @aluzzardi will be very cool and useful, but it will *not* be the same outcome as what I'm describing IMO.

**Reply by marcosnils** [reply]
Yep, agree with all this. It's one of the possibilities that we spoke the other day with @aluzzardi and I particularly like that this is extremely useful in the context of improving adoption. I think we could take some inspiration from the cloud native buildpacks (https://buildpacks.io/) proejct to understand how they're currently doing this "automatic" runtime detection / configuration since they do something very similar.

**Reply by shykes** [reply]
See also #24 for a continuation of my thoughts on this :)

**Comment by sipsma** [comment]
One quick note to self and @aluzzardi about implementation details: this goal seems like it will require that we support subresolvers in user actions (i.e. fields not directly under `Query` can also be explicit, custom written resolvers). We weren't sure if we wanted to support those initially just because it's more to document for developers, but this all seems like a compelling enough use case that we should. Luckily it won't be a huge refactor or anything like that, just need to add a few extra blocks of code here and there to handle the extra cases introduced by this.

**Reply by aluzzardi** [reply]
Yep.

Related but I was thinking we should stop doing the "query mangling" ourselves and instead rely on the schema doing an `extend Query { yarn: Yarn! } }` instead. Relies on a native way to do it, less magic, opens the door for `extend Filesystem { yarn: Yarn! }`

That however requires sub resolvers already

**Reply by sipsma** [reply]
> Related but I was thinking we should stop doing the "query mangling" ourselves and instead rely on the schema

Yes 100%, the mangling is another hacky artifact of the first draft of the switch to graphql. `extend` is the much better approach.

**Comment by dolanor** [comment]
Really like this idea.
It would be very compelling. I don't have any opinions on the technical side because I didn't dig in it enough to understand all of what you're talking about.

I guess specific tasks wouldn't work out of the box:
- makefile spinning up local server as dependencies
- docker-compose running e2e testing or setting up deps (does buildkit support running docker services inside it?)
- specific environment settings that wouldn't propagate in dagger by default
